File: apps/mobile/lib/data/simulation/simulation_engine.dart
import 'dart:async';

import '../../core/utils/geo.dart';
import '../mock/demo_data.dart';
import '../models/stop.dart';

/// Simülasyon anlık durumu.
class SimulationTick {
  const SimulationTick({
    required this.location,
    required this.heading,
    required this.speedKmh,
    required this.nextStopIndex,
    required this.atStopIndex,
    required this.finished,
  });

  final LatLngPoint location;
  final double heading;
  final double speedKmh;
  final int nextStopIndex;
  final int? atStopIndex;
  final bool finished;
}

/// Backend olmadan aracın harita üzerinde hareketini simüle eder.
/// Konumu günceller, geçilen durakları tespit eder ve dinleyicilere yayınlar.
/// WebSocket geldiğinde bu motor gerçek konum akışıyla değiştirilir.
class SimulationEngine {
  SimulationEngine({
    required this.path,
    required this.stops,
    this.interval = const Duration(seconds: 2),
    this.baseSpeedKmh = 34,
    int startIndex = 0,
  }) : _index = startIndex;

  final List<LatLngPoint> path;
  final List<Stop> stops;
  final Duration interval;
  final double baseSpeedKmh;

  int _index;
  Timer? _timer;
  final _controller = StreamController<SimulationTick>.broadcast();

  Stream<SimulationTick> get stream => _controller.stream;
  bool get isRunning => _timer != null;

  /// Mevcut anlık durum.
  SimulationTick get current => _buildTick();

  void start() {
    if (_timer != null) return;
    _emit(_buildTick());
    _timer = Timer.periodic(interval, (_) => _step());
  }

  void stop() {
    _timer?.cancel();
    _timer = null;
  }

  void dispose() {
    stop();
    _controller.close();
  }

  void _step() {
    if (_index >= path.length - 1) {
      stop();
      _emit(_buildTick(finished: true));
      return;
    }
    _index++;
    _emit(_buildTick());
  }

  LatLngPoint get _location => path[_index.clamp(0, path.length - 1)];

  int _computeNextStopIndex() {
    for (var i = 0; i < stops.length; i++) {
      if (!_hasPassedStop(i)) return i;
      if (isWithinRadius(_location, stops[i].point, stops[i].radiusMeters.toDouble())) return i;
    }
    return stops.length - 1;
  }

  bool _hasPassedStop(int stopIndex) => _index > stopIndex * simulationSegmentsPerStop;

  int? _atStop() {
    for (var i = 0; i < stops.length; i++) {
      if (isWithinRadius(_location, stops[i].point, stops[i].radiusMeters.toDouble())) return i;
    }
    return null;
  }

  SimulationTick _buildTick({bool finished = false}) {
    final loc = _location;
    final prev = path[(_index - 1).clamp(0, path.length - 1)];
    final moving = haversineMeters(prev, loc) > 1;
    return SimulationTick(
      location: loc,
      heading: bearingDeg(prev, loc),
      speedKmh: moving ? baseSpeedKmh : 0,
      nextStopIndex: _computeNextStopIndex(),
      atStopIndex: _atStop(),
      finished: finished || _index >= path.length - 1,
    );
  }

  void _emit(SimulationTick tick) {
    if (!_controller.isClosed) _controller.add(tick);
  }
}

File: apps/mobile-legacy-rn/src/services/simulation/SimulationEngine.ts
import type { LatLng, Stop } from '@servis/shared';
import { bearingDeg, haversineMeters, isWithinRadius } from '@/lib/geo';

export interface SimulationTick {
  /** Aracın güncel konumu. */
  location: LatLng;
  /** Yön (derece). */
  heading: number;
  /** Anlık hız (km/s). */
  speedKmh: number;
  /** Sıradaki durağın index'i. */
  nextStopIndex: number;
  /** Araç şu an bir durağın yarıçapı içinde mi? */
  atStopIndex: number | null;
  /** Yol tamamlandı mı? */
  finished: boolean;
}

export type SimulationListener = (tick: SimulationTick) => void;

interface SimulationOptions {
  path: LatLng[];
  stops: Stop[];
  /** Kaç ms'de bir adım ilerlesin. */
  intervalMs?: number;
  /** Ortalama hız (görsel amaçlı). */
  speedKmh?: number;
  /** Başlangıç path index'i. */
  startIndex?: number;
}

/**
 * Backend olmadan aracın harita üzerinde hareketini simüle eder.
 * Konumu günceller, geçilen durakları tespit eder ve dinleyicilere yayınlar.
 * WebSocket geldiğinde bu motor gerçek konum akışıyla değiştirilir.
 */
export class SimulationEngine {
  private readonly path: LatLng[];
  private readonly stops: Stop[];
  private readonly intervalMs: number;
  private readonly baseSpeedKmh: number;
  private index: number;
  private timer: ReturnType<typeof setInterval> | null = null;
  private listeners = new Set<SimulationListener>();

  constructor(opts: SimulationOptions) {
    this.path = opts.path;
    this.stops = opts.stops;
    this.intervalMs = opts.intervalMs ?? 2000;
    this.baseSpeedKmh = opts.speedKmh ?? 34;
    this.index = opts.startIndex ?? 0;
  }

  subscribe(listener: SimulationListener): () => void {
    this.listeners.add(listener);
    // İlk durumu hemen ilet
    listener(this.buildTick());
    return () => this.listeners.delete(listener);
  }

  start(): void {
    if (this.timer) return;
    this.timer = setInterval(() => this.step(), this.intervalMs);
  }

  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  isRunning(): boolean {
    return this.timer !== null;
  }

  private step(): void {
    if (this.index >= this.path.length - 1) {
      this.stop();
      this.emit(this.buildTick(true));
      return;
    }
    this.index += 1;
    this.emit(this.buildTick());
  }

  private currentLocation(): LatLng {
    return this.path[Math.min(this.index, this.path.length - 1)]!;
  }

  private computeNextStopIndex(loc: LatLng): number {
    // Araca en yakın "ileri" durak: yol boyunca sırayla ilk ulaşılmamış durak.
    for (let i = 0; i < this.stops.length; i++) {
      const stop = this.stops[i]!;
      const passed = this.hasPassedStop(i);
      if (!passed) return i;
      if (isWithinRadius(loc, stop, stop.radius_meters)) return i;
    }
    return this.stops.length - 1;
  }

  private hasPassedStop(stopIndex: number): boolean {
    // Path, duraklar arasına eşit bölünmüş (buildPath: 8 segment/durak arası).
    const segmentsPerStop = 8;
    const stopPathIndex = stopIndex * segmentsPerStop;
    return this.index > stopPathIndex;
  }

  private atStop(loc: LatLng): number | null {
    for (let i = 0; i < this.stops.length; i++) {
      const stop = this.stops[i]!;
      if (isWithinRadius(loc, stop, stop.radius_meters)) return i;
    }
    return null;
  }

  private buildTick(finished = false): SimulationTick {
    const loc = this.currentLocation();
    const prev = this.path[Math.max(0, this.index - 1)]!;
    const heading = bearingDeg(prev, loc);
    const moving = haversineMeters(prev, loc) > 1;
    return {
      location: loc,
      heading,
      speedKmh: moving ? this.baseSpeedKmh : 0,
      nextStopIndex: this.computeNextStopIndex(loc),
      atStopIndex: this.atStop(loc),
      finished: finished || this.index >= this.path.length - 1,
    };
  }

  private emit(tick: SimulationTick): void {
    this.listeners.forEach((l) => l(tick));
  }
}
